File: go/0347M-前K个高频元素.go
package main

import "container/heap"

type IntHeap [][2]int // 0 key 1 value

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i][1] < h[j][1] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	*h = append(*h, x.([2]int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func topKFrequent(nums []int, k int) []int {
	m := make(map[int]int)
	ans := make([]int, k)
	h := &IntHeap{}
	heap.Init(h)
	for _, v := range nums {
		m[v]++
	}
	for key, value := range m {
		heap.Push(h, [2]int{key, value})
		if h.Len() > k {
			heap.Pop(h)
		}
	}
	for k > 0 {
		k--
		ans[k] = heap.Pop(h).([2]int)[0]
	}
	return ans
}

File: cpp/0347M-前 K 个高频元素.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-20 14:42:05
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    struct cmp {
        bool operator()(const pair<int, int>& t1, const pair<int, int>& t2) {
            return t1.second > t2.second;
        }
    };

    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        for (int num : nums) mp[num]++;
        priority_queue<pair<int, int>, vector<pair<int, int>>, cmp> que;
        for (auto it = mp.begin(); it != mp.end(); it++) {
            que.push(*it);
            if (que.size() > k) que.pop();
        }
        vector<int> ans;
        while (k--) {
            int t = que.top().first;
            que.pop();
            ans.push_back(t);
        }
        return ans;
    }
};